cmd/erasure-object.go

```diff
@@ -1497,12 +1511,18 @@ func (er erasureObjects) DeleteObject(ctx context.Context, bucket, object string
 			// any version specified explicitly. Or if a particular
 			// version id needs to be replicated.
 			if err = er.deleteObjectVersion(ctx, bucket, object, writeQuorum, fi, opts.DeleteMarker); err != nil {
-				return objInfo, toObjectErr(err, bucket, object)
+				// soft deletes are idempotent. Return the topmost delete marker in stack if delete marker already set
+				if !errors.As(err, &errDeleteNoOp) {
+					return objInfo, toObjectErr(err, bucket, object)
+				}
+				oi, err := er.getObjectInfo(ctx, bucket, object, ObjectOptions{})
+				if err != nil && oi.DeleteMarker {
+					fi.VersionID = oi.VersionID
+				}
```

Comment:
let deleteObjectVersion - return the top level `oi` instead of calling getObjectInfo again.